src/main/java/com/thealgorithms/datastructures/graphs/BoruvkaAlgorithm.java

```diff
@@ -0,0 +1,217 @@
+ */
+
+final class BoruvkaAlgorithm {
+    private BoruvkaAlgorithm() {
+    }
+
+    /**
+     * Represents an edge in the graph
+     */
+    static class Edge {
+        final int src;
+        final int dest;
+        final int weight;
+
+        Edge(final int src, final int dest, final int weight) {
+            this.src = src;
+            this.dest = dest;
+            this.weight = weight;
+        }
+    }
+
+    /**
+     * Represents the graph
+     */
+    static class Graph {
+        final int vertex;
+        final List<Edge> edges;
+
+        /**
+         * Constructor for the graph
+         *
+         * @param vertex number of vertices
+         * @param edges  list of edges
+         */
+        Graph(final int vertex, final List<Edge> edges) {
+            if (vertex < 0) {
+                throw new IllegalArgumentException("Number of vertices must be positive");
+            }
+            if (edges == null || edges.isEmpty()) {
+                throw new IllegalArgumentException("Edges list must not be null or empty");
+            }
+            for (final var edge : edges) {
+                checkEdgeVertices(edge.src, vertex);
+                checkEdgeVertices(edge.dest, vertex);
+            }
+
+            this.vertex = vertex;
+            this.edges = edges;
+        }
+    }
+
+    /**
+     * Represents a subset for Union-Find operations
+     */
+    static class Component {
+        int parent;
+        int rank;
+
+        Component(final int parent, final int rank) {
+            this.parent = parent;
+            this.rank = rank;
+        }
+    }
+
+    /**
+     * Represents the state of Union-Find components and the result list
+     */
+    static class BoruvkaState {
+        List<Edge> result;
+        Component[] components;
+        Graph graph;
+
+        BoruvkaState(Graph graph) {
+            this.result = new ArrayList<>();
+            this.components = initializeSubsets(graph);
+            this.graph = graph;
+        }
+
+        /**
+         * Adds the cheapest edges to the result list and performs Union operation on the subsets.
+         *
+         * @param cheapest Array containing the cheapest edge for each subset.
+         */
+        void merge(Edge[] cheapest) {
+            for (int i = 0; i < graph.vertex; ++i) {
+                if (cheapest[i] != null) {
+                    final var set1 = find(components, cheapest[i].src);
+                    final var set2 = find(components, cheapest[i].dest);
+
+                    if (set1 != set2) {
+                        result.add(cheapest[i]);
+                        union(components, set1, set2);
+                    }
+                }
+            }
+        }
+
+        /**
+         * Checks if there are more edges to add to the result list
+         *
+         * @return true if there are more edges to add, false otherwise
+         */
+        boolean hasMoreEdgesToAdd() {
```

Comment:
Maybe a better name would be `isReady()`?

Reply:
I think `hasMoreEdgesToAdd()` is more precise.

Reply:
You use it in the line `boruvkaState.hasMoreEdgesToAdd()`. My first impression would be that _you add edges to the boruvkaState_, which is not entirely true. Maybe `isMST()`?

Reply:
Renaming the method to `isMST()` might be misleading because it doesn't directly check if the MST is complete. Instead, it checks if there are more edges to add to the MST.